File: sort/Insertion Sort/src.py
"""
ALGORITHM : Insertion Sort
WORST CASE => {
    PERFORMANCE:  O(n^2)
    SPACE:        O(1)
}
"""


def insertion_sort(arr):
    for k in range(1, len(arr)):
        for i in range(k, 0, -1):
            # This Block runs `n(n+1)/2` times hence T(N^2)
            if arr[i - 1] > arr[i]:
                arr[i], arr[i - 1] = [arr[i - 1], arr[i]]

    return arr


if __name__ == '__main__':
    array = [8, 7, 6, 5, 4, 3, 2, 1]
    sorted_arr = insertion_sort(array)
    print(sorted_arr)

File: Algorithms/Sort/Insertion Sort/src.java
/*
ALGORITHM : Insertion Sort
WORST CASE => {
    PERFORMANCE:  O(n^2)
    SPACE:        O(1)
}
*/

import java.util.Arrays;

public class InsertionSort {
    public static void main(String[] args)  {
        int[] arr;
        arr = new int[] {6, 5, 4, 3, 2, 1};

        int l = arr.length;
        int temp;
        for(int i=1;i<l;i++) {
            for(int j=i;j>0;j--) {
                if (arr[j-1] > arr[j]) {
                    temp = arr[j-1];
                    arr[j-1] = arr[j];
                    arr[j] = temp;
                }
            }
        }

        System.out.println(Arrays.toString(arr));
    }
}
